[MRG] Fast PolynomialFeatures on CSR matrices

AWNystrom opened the pull request:
This PR allows PolynomialFeatures to operate on compressed sparse row (CSR) matrices. It does so in a way that only takes products of nonzero features, so it scales with the density of the matrix to the power of the degree of the expansion.

Here's a plot that compares the runtime of this method to the dense and crc methods: 
![image](https://user-images.githubusercontent.com/1478067/47197120-06692f80-d319-11e8-8494-a5146a77a6aa.png)




See https://arxiv.org/pdf/1803.06418.pdf for details of the algorithm.

AWNystrom committed:
csr polynomial expansions are now possible

AWNystrom committed:
Adding a prefix to csr_expansion.pyx

AWNystrom committed:
Adding a description to the docstring.

AWNystrom committed:
Shameless plug in docstring

AWNystrom committed:
so close. Something's off

AWNystrom committed:
need to clean up

AWNystrom committed:
ready to go

AWNystrom committed:
cleaned up imports

AWNystrom committed:
pep8

AWNystrom committed:
more pep8

ogrisel reviewed (commented):
Thanks for your contribution, here is a first batch of comments.

Also the build is broken under both linux and windows but I have not investigated. The cython module cannot be imported.

Could also please include the results of the benchmarks in a comment with your analysis to help other reviewer without having them to locally build and run the benchmark of this PR?

ogrisel commented on benchmarks/bench_feature_expansions.py:
```diff
@@ -0,0 +1,52 @@
+        for dim_index, dim in enumerate(dimensionalities):
+            print(trial, density, dim)
+            X_csr = sparse.random(num_rows, dim, density).tocsr()
+            X_dense = X_csr.toarray()
```
Also could you please include `X_csc = X_csr.tocsc()` in this benchmark?

AWNystrom replied:
Sure, but csc is crazy slow, so I'm adding a subplot where I compare csc and csr. Otherwise dense and csr would be almost indistinguishable because csc's times would squash everything. Sound OK?

ogrisel replied:
Maybe use a log scale (and do not compute CSC for the largest problem to keep the overall runtime of the benchmark under control).

AWNystrom replied:
How's that looking?

https://drive.google.com/file/d/1IUZSsP_mdz5UkVx-Mfi4kte2Jd0yT-Bo/view

ogrisel replied:
Thanks! Note that it's possible to directly report plot in the github discussion:

![image](https://user-images.githubusercontent.com/89061/46336439-14d1fe80-c62b-11e8-8d3f-06d06b4c1762.png)

Maybe we could have done one sub plot per density level but this is nitpicking. I think the point is that the new CSR-based implementation is significantly faster than both the dense and the CSC variants.

AWNystrom replied:
Interesting difference between our machines.

I did implement a Cython version for dense when I wrote the paper, just so that the comparison would be fair. I found that dense started off faster, but as D grew and d shrank, sparse became faster, as one would expect.

AWNystrom replied:
Here's the loglog plot:
![image](https://user-images.githubusercontent.com/1478067/46395169-6072bf80-c6a0-11e8-80cd-63b76ef40e4c.png)

AWNystrom replied:
Regarding separating the plots into subplots per density, since only the csr method varies with density, each subplot would be pretty similar. Having them all on the same plot with an alpha < 1 kind of shows that only the csr method varies with density, which is partially what I wanted to show. Let me know if you want the change though.

ogrisel replied:
Sorry I had not seen these comments. I think the fact that it was faster than dense was because we ran benchmarks on a very low number of samples which was not representative of realistic situation. The new setting is more interesting. I think.

AWNystrom committed:
round 1 of ogrisel review

AWNystrom committed:
I was too hasty with the tests. Fixed.

AWNystrom committed:
5 trials in benchmark, not 1

AWNystrom committed:
updating dimensionalities in benchmark

AWNystrom committed:
pep8 and trying to get scipy.sparse.random to work on Travis

ogrisel commented on benchmarks/bench_feature_expansions.py:
```diff
@@ -0,0 +1,70 @@
+trials = 5
+num_rows = 100
+dimensionalities = [50, 100, 200, 400, 800]
+densities = [0.1, 0.4, 0.7, 1.0]
```
I think it would be more interesting to explore densities such as  `[0.01, 0.1, 1]`. I think sparse representations only make sense below 0.1 density.

ogrisel commented on sklearn/preprocessing/data.py:
```diff
@@ -1444,40 +1445,65 @@ def transform(self, X):
-            XP = sparse.hstack(columns, dtype=X.dtype).tocsc()
+        if sparse.isspmatrix_csr(X):
+            if self.degree > 3:
+                return (self.transform(X.tocsc())).tocsr()
```
I would not call `.tocsr()` back at the end in this case. Let's the next step in the estimator pipeline do another conversion itself only if necessary.

AWNystrom replied:
It would seem odd to me to get a csr matrix returned for some degrees, but get a crc matrix returned for others. It seems this could mess up some people's pipelines. If you feel I shouldn't convert to the same format the user input, I'm happy to change this, but it seems a bit odd to me.

ogrisel replied:
scikit-learn estimators automatically convert sparse matrix format when then need so I think it's fine not to convert a priori and spare a potentially useless memory copy, although I agree this is not a big deal.

AWNystrom replied:
The csc PolynomialFeatures test checked that the output was also a csc matrix, so I figured I'd follow the same pattern. Removing this conversion would make that part of the test fail. If it's really OK for the output to not be in the same format as the input, I'll go ahead and remove it. I thought that some algorithms ran much slower on certain sparse formats, so it seems users might expect this consistency though. Let me know.

ogrisel commented on sklearn/preprocessing/data.py:
```diff
@@ -34,8 +34,9 @@
                                 FLOAT_DTYPES)
 
-from ._encoders import OneHotEncoder
+from ._csr_expansion import _csr_expansion
```
Please rename both the module and the function to `_csr_polynomial_expansion` to make it more explicit.

AWNystrom committed:
Changing benchmark plot, changing _csr_expansion to _csr_polynomial_expansion

AWNystrom committed:
Now works when not enough dims for an expansion up to the full degree, but enough for a lower degree.

AWNystrom commented:
I seem to have made tests fail with the last push. Sorry about that. I'll fix it when I get a chance.

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 8e450a5289181cf12f96bc18ecf9d9641cca0487 into acf3babfe5aa16e3189f1d2f7d9ffe45357630a8 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-2ba3bfd94b894c9c5c157957750c5de1cc838606)

**new alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AWNystrom committed:
correcting min_required_dim

ogrisel commented:
Please delete the `sklearn/preprocessing/csr_expansion.pyx` file that is no longer used.

ogrisel commented:
There is a travis job configuration that fails with:

```
from scipy.sparse import random as sparse_random
ImportError: cannot import name random
```

However for scikit-learn 0.21 we will bump up the scipy dependency to 0.17 (#12184). So in the mean time have you can do:

```python
try:
    from scipy.sparse import random as sparse_random
except ImportError:
    # Avoid import error when running the tests with scipy < 0.17
    sparse_random = None
...

# TODO: remove skipif once scipy < 0.17 is no longer supported
@pytest.mark.skipif(sparse_random is None)
def test_that_needs_sparse_random():
   ...
```

ogrisel commented:
The fact that the CSR expansion is still significantly faster than the dense version even when density == 1.0 in the benchmark makes me think that the dense case could also receive some love but this should probably be done in a separate PR once this one is merged.

ogrisel commented on benchmarks/bench_feature_expansions.py:
```diff
@@ -0,0 +1,61 @@
+degree = 2
+trials = 5
+num_rows = 100
+dimensionalities = np.array([50, 100, 150, 200])
```
Maybe we could instead explore:

```python
dimensionalities = np.array([25, 50, 100, 200])
```

and then use a loglog plot so that straight lines would reflect linear time complexities and would make it easier to interpret the plots.

AWNystrom committed:
Implementing sparse_random so tests pass on old configs

AWNystrom committed:
Updated dimensionalities and logscale on both x and y axes

AWNystrom committed:
Setting trials back to 5

AWNystrom committed:
Using scipy's sparse.random when we can

AWNystrom committed:
Merge branch 'master' into csr_expansion

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 06d86bf0bff505051c7b2ba771cf84cd740d9132 into f456a4082879656c3485adde3d12c143b8dc6c3a - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-859722c7a1e7ce3101847f464cca2b4ce8885fd2)

**new alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AWNystrom committed:
Fixing my bad merge

AWNystrom committed:
Trying to get travis to not think this except is blank

AWNystrom committed:
Update test_data.py so not have an empty except

ogrisel committed:
Refactory cython expansion to release the GIL

ogrisel committed:
Create an isolated RNG instance for the sparse tests

ogrisel committed:
Simplify the CSR return value

ogrisel commented:
Thanks @AWNystrom. I just pushed a small refactoring to release the GIL in the CPU-intensive loop. It should not have degraded the perf or the results. I also simplified the construction of the csr matrix and the changed the seeding of the RNG in the tests.

Let me know if you have any comment. For the future, we might want to also use a `prange` outer loop on the rows we should leave that for another PR. (note that #11950 is already experimenting with the use of openmp and prange for the k-means loops).

ogrisel committed:
Fix tests for old scipy

ogrisel committed:
Tweaked benchmark

ogrisel committed:
More simplification of the benchmark plots

ogrisel committed:
Favor the CSR representation over the slow csc

ogrisel commented:
> One thing I was considering in _csr_polynomial_expansion is separating out the terms that only involve i, i and j or j, and k, and caching them in each loop so that they don't have to be recomputed each time. That might give another speedup, but it would make the code less readable, which I'm guessing conflicts with scikit-learn's principles.

Good idea. This is a trade of it depends on the speed up :) If we get a 2x speed up and the code with caching, it's probably fine to have slightly more complex code. Have you tried to implement that caching strategy already and benchmark it?

I assume the increase memory usage would be negligible compared to the size of the output CSR matrix if n_samples is 100 or more.

ogrisel reviewed (approved):
I am ok with merging this PR in its current state.

The scikit-learn policy is to not implement too recent papers but I think in this case it's fine as:

- the code is not too complex (although the math of the paper is not that trivial ;);
- there is a very clear performance gain;
- this is only about a pure computational performance gain: there is no speed vs accuracy tradeoff. The output of the transformer is the same as previously, it's just significantly faster and more scalable on sparse data.

The caching might be implemented here or in a subsequent PR, as other reviewers might prefer.

ogrisel commented:
BTW I am curious @AWNystrom: for each kind of data science pipeline did you find sparse polynomial features most useful?

- text bag of words vectorization => polynomial expansion => classifier?
- categorical variable =>  one hot encoding  => polynomial expansion => classifier?
- numerical features => binning/quantization => one hot encoding => polynomial expansion => classifier?

For the text case, the data is so high dim in the beginning that it's probably worthless to do polynomial feature extensions. The latter two cases are more promising.

It would be great to come up with an example case on a typical benchmark dataset from `sklearn.datasets` where the usefulness of polynomial features can be demonstrated.

ogrisel commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -0,0 +1,159 @@
+def _csr_polynomial_expansion(ndarray[DATA_T, ndim=1] data,
+                              ndarray[INDEX_T, ndim=1] indices,
+                              ndarray[INDEX_T, ndim=1] indptr,
+                              int has_sorted_indices, INDEX_T d,
```
I just noticed that I left an unused parameter in my refactoring. I think this `has_sorted_indices` flag  is no longer required. The `csr_matrix` constructor we now use should automatically set this attribute I think.

ogrisel commented on sklearn/preprocessing/data.py:
```diff
@@ -1452,41 +1453,69 @@ def transform(self, X):
+            to_stack.append(X)
+            for deg in range(2, self.degree+1):
+                min_required_dim = deg + int(self.interaction_only) - 1
+                if X.shape[1] < min_required_dim:
```
Indeed, good catch. I had not reviewed correctly that section of the PR.

AWNystrom commented:
Hey thanks for pointing out this issue. I have family visiting for the next
few days so I’ll not be able to fix it right away, but will do so ASAP.
On Fri, Oct 5, 2018 at 1:11 PM Olivier Grisel <notifications@github.com>
wrote:

> *@ogrisel* commented on this pull request.
> ------------------------------
>
> In sklearn/preprocessing/data.py
> <https://github.com/scikit-learn/scikit-learn/pull/12197#discussion_r223126209>
> :
>
> > -                    for col_idx in comb:
> -                        out_col = X[:, col_idx].multiply(out_col)
> -                    columns.append(out_col)
> -                else:
> -                    columns.append(sparse.csc_matrix(np.ones((X.shape[0], 1))))
> -            XP = sparse.hstack(columns, dtype=X.dtype).tocsc()
> +        if sparse.isspmatrix_csr(X):
> +            if self.degree > 3:
> +                return self.transform(X.tocsc()).tocsr()
> +            to_stack = []
> +            if self.include_bias:
> +                to_stack.append(np.ones(shape=(n_samples, 1), dtype=X.dtype))
> +            to_stack.append(X)
> +            for deg in range(2, self.degree+1):
> +                min_required_dim = deg + int(self.interaction_only) - 1
> +                if X.shape[1] < min_required_dim:
>
> Indeed, good catch. I had not reviewed correctly that section of the PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/12197#discussion_r223126209>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABaNs7pOdnjW8hZJYDWFywpjzsIW6LLdks5uh70LgaJpZM4XAW5G>
> .
>

AWNystrom committed:
Fixing dimensionality edge cases

AWNystrom commented:
> BTW I am curious @AWNystrom: for which kind of data science pipeline did you find sparse polynomial features most useful?
> 
> * bag of words text vectorization => polynomial expansion => classifier?
> * categorical variables =>  one hot encoding  => polynomial expansion => classifier?
> * numerical features => binning/quantization => one hot encoding => polynomial expansion => classifier?
> 
> For the text case, the data is so high dim in the beginning that it's probably worthless to do polynomial feature extensions. The latter two cases are more promising.
> 
> It would be great to come up with an example case on a typical benchmark dataset from `sklearn.datasets` where the usefulness of polynomial features can be demonstrated.

Olivier, the reason worked on figuring out how to solve this problem is this:

I was working on computing lifetime expected value of customers given demographics and spending history. The data was sparse, but not of terribly high dimensionality. I suspected that there might be some "this feature AND that feature" relations that were important, and the results had to be easily interpretable, so feature products seemed like a good place to start. I therefore tried passing my CSR matrix to PolynomialFeatures, and of course, got an error. I was then driven to figure out how they could be computed on a CSR matrix, and here we are :)

AWNystrom commented:
> > BTW I am curious @AWNystrom: for which kind of data science pipeline did you find sparse polynomial features most useful?
> > 
> > * bag of words text vectorization => polynomial expansion => classifier?
> > * categorical variables =>  one hot encoding  => polynomial expansion => classifier?
> > * numerical features => binning/quantization => one hot encoding => polynomial expansion => classifier?
> > 
> > For the text case, the data is so high dim in the beginning that it's probably worthless to do polynomial feature extensions. The latter two cases are more promising.
> > It would be great to come up with an example case on a typical benchmark dataset from `sklearn.datasets` where the usefulness of polynomial features can be demonstrated.
> 
> Olivier, the reason worked on figuring out how to solve this problem is this:
> 
> I was working on computing lifetime expected value of customers given demographics and spending history. The data was sparse, but not of terribly high dimensionality. I suspected that there might be some "this feature AND that feature" relations that were important, and the results had to be easily interpretable, so feature products seemed like a good place to start. I therefore tried passing my CSR matrix to PolynomialFeatures, and of course, got an error. I was then driven to figure out how they could be computed on a CSR matrix, and here we are :)

Oh but I didn't quite specify which of the three types of pipelines that problem was. It was both 2 and 3. I tried applying the degree=2 case to a TF-IDF 20 Newsgroups CSR matrix, but the resulting dimensionality was larger than an int32, which is the type scipy uses for indices. For kicks I built a CSR expansion using int64 indices, and I believe it completed in about 20 minutes on my laptop, which I thought was pretty good considering the insane dimensionality.

ogrisel commented:
Thanks for the feedback.

Maybe we should consider using int64 for the indices of the CSR matrix returned by this Cython code.

ogrisel commented:
Thanks for the edge case fix. @TomDLT anything else to add?

TomDLT reviewed (approved):
I did not review the Cython code, but the tests and benchmark are convincing.
LGTM

ogrisel commented:
@AWNystrom please add an entry to document this in `doc/whats_new/v0.21.rst`.

AWNystrom committed:
Getting rid of has_sorted_indices in the docstring

AWNystrom committed:
adding entry in whats_new

AWNystrom committed:
Merge branch 'master' into csr_expansion

ogrisel commented:
Thanks @AWNystrom for your quick reaction time!

@jnothman @amueller @GaelVaroquaux @agramfort I would love to get a +3 for this PR given the fact that it's borderline without our new method policy in scikit-learn: https://github.com/scikit-learn/scikit-learn/pull/12197#pullrequestreview-161968430

GaelVaroquaux commented:
Given that it is about speed up a lot a classic problem, I am not completely against it. However, for the sake of principles, can we at least wait for the paper to be published. We do need to put a bit of safeguards.

ogrisel commented:
@AWNystrom did you submit your manuscript to a peer-reviewed journal / conference or do you treat it as a technical report?

AWNystrom commented:
> @AWNystrom did you submit your manuscript to a peer-reviewed journal / conference or do you treat it as a technical report?

I tried getting it into ICML, but I got the impression they were after deep learning papers, not improvements on an classic methods like this. Since then I've been treating it as a technical report.

GaelVaroquaux commented:
> I tried getting it into ICML, but I got the impression they were after deep
> learning papers, not improvements on an classic methods like this. Since then
> I've been treating it as a technical report.

Hum, makes sense.

It's a bit of an edge case in our decision process. I think that this
warrants a vote.

jnothman commented:
> it's borderline w.r.t. our recent paper implementation policy in scikit-learn

Well I think the "mature algorithm" policy serves two purposes:
1. to avoid maintaining API for several alternative implementations for similar purposes, as each gets trumped by a newcomer.
2. to make it more likely that core or contributing devs can confidently maintain the code and ensure its correctness.

Efficiency improvements do not tend to affect API, and so the first purpose doesn't apply. We always can choose to revert the inclusion, without the risk of breaking user code's functionality.

In this case, I think the algorithm is relatively straightforward to understand wrt correctness, and other reviewers have lauded the tests, so I don't think it should be held back by that policy.

[I might note that some aspects of this algorithm (I've only skimmed it) remind me of the CSR-CSR matrix multiplication algorithm. One implementation we had considered for sparse PolynomialFeatures was to use matrix multiplication with the log-sum-exp trick... This obviously has several advantages over that.]

ogrisel commented:
I agree with @jnothman's analysis. Let's wait a bit for the other pinged people to voice their opinion though.

jnothman commented:
Maybe it's not so similar to CSR-CSR. Either way it was musing rather than a request for changes...

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -55,6 +55,13 @@ Support for Python 3.4 and below has been officially dropped.
   of calculating it every time on the fly.
   :issue:`12116` by :user:`Ekaterina Krivich <kiote>` and `Joel Nothman`_.
 
+- |Feature| |Efficiency| :class:`preprocessing.PolynomialFeatures` now supports
```
I don't think we should say feature here. It is not a new capability in the software

jnothman reviewed (commented):
I'd make that change but I'm on the move

AWNystrom commented:
> Maybe it's not so similar to CSR-CSR. Either way it was musing rather than a request for changes...

No worries. I was just afraid I'd missed an important connection.

AWNystrom committed:
Update v0.21.rst

jnothman commented:
(and I don't know how Ben fits in the picture...!)

AWNystrom commented:
> (and I don't know how Ben fits in the picture...!)

(I told him I'd try to weave that in, and that was my attempt.)

AWNystrom commented:
Thoughts on logging a warning when PolynomialFeatures is called on a CSC matrix with degree 2 or 3 that recommends using a CSR?

jnothman commented:
Ambivalent? We don't currently warn on CSR-CSC conversion e.g. in random forests, but perhaps this is different. Scipy has SparseEfficiencyWarnings, but they usually indicate something different to this.

AWNystrom commented:
> Ambivalent? We don't currently warn on CSR-CSC conversion e.g. in random forests, but perhaps this is different. Scipy has SparseEfficiencyWarnings, but they usually indicate something different to this.

Whatever you think is best. Alternatively, we could do a return self.transform(X.tocsr()).tocsc() in this case. The CSC is just so slow that the user would benefit from warning or automatic conversion to CSR. The latter would double the memory used for holding the data, so that's an argument against it.

AWNystrom committed:
Update data.py

sklearn-lgtm commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 876ac2f30a9e5d387b38c18351b9ed96da7fba3a into bea1eb51b34d5d61a3df217f5b7ed25f6b98eec3 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-021589fe9e6b8fd56aebca124318c0421b14d265)

**new alerts:**

* 1 for Syntax error

**fixed alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AWNystrom committed:
Update data.py

AWNystrom committed:
Update test_data.py

Since we call the CSR method when able given a CSC input, adding cases where the CSR method has to be used because the degree's too high.

AWNystrom commented:
OK, I can't think of anything else that should be done. Let me know if there's anything else you'd like changed.

AWNystrom commented:
Actually, doing this conversion to CSR on CSC input invalidates the benchmark. We could add a parameter to the PolynomialFeatures constructor that prevents the conversion.

jnothman commented:
It invalidates the benchmark meaning that it invalidates the benchmark
script, or it means that this is no longer the most efficient?

AWNystrom commented:
The method in the benchmark labeled “crc” is actually the CSR method
because CSCs get converted to CSRs and use the CSR method when the defree’s
2 or 3. This makes the CSC method look fast in the benchmark because it’s
not actually the CSC method. How about we just remove the benchmark?
On Wed, Oct 17, 2018 at 7:19 PM Joel Nothman <notifications@github.com>
wrote:

> It invalidates the benchmark meaning that it invalidates the benchmark
> script, or it means that this is no longer the most efficient?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/12197#issuecomment-430851988>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABaNs_AoS60935CZ005RuGhCTIJfkmJlks5ul-U1gaJpZM4XAW5G>
> .
>

jnothman commented:
Yes, remove the CSC benchmark

ogrisel commented:
And because the new benchmark will only run without CSC we can change it to run for larger number of features.

ogrisel commented:
Or alternatively without the log scale to make the speed-up effect look more dramatic ;)

AWNystrom committed:
Removing csr from the benchmark, no longer logscale

AWNystrom committed:
Removing csc from the benchmark, no longer logscale

AWNystrom committed:
Merge branch 'csr_expansion' of https://github.com/AWNystrom/scikit-learn into csr_expansion

AWNystrom commented:
It's been done. Plot updated at the top of this page.

ogrisel commented:
Merged! Thank you very much @AWNystrom!

AWNystrom commented:
> Merged! Thank you very much @AWNystrom!

Thank you! It's been a pleasure contributing. Thanks for the code reviews, everyone!
